pontoon/pretranslation/transformer.py

```diff
@@ -0,0 +1,238 @@
+        return FTL.TextElement(f'<span id="pt-{idx}" translate="no">{clean}</span>')
+
+
+class PreparePretranslation(Transformer):
+    """
+    Flattens the given Pattern, uplifting selectors to the highest possible level and
+    duplicating shared parts in the variants. All other Placeables are serialised as
+    TextElements.
+    """
+
+    def __init__(self, locale: Locale):
+        self.locale = locale
+        self.replacements: list[str] = []
+
+    def applyReplacements(self, translation: str):
+        return re.sub(
+            r'<span id="pt-(\d+)" translate="no">[^<>]*</span>',
+            lambda m: self.replacements[int(m.group(1))],
+            translation,
+        )
+
+    def visit_Attribute(self, node: FTL.Attribute):
+        flatten_select_expressions(node.value)
+        return self.generic_visit(node)
+
+    def visit_Message(self, node: FTL.Message):
+        if node.value:
+            flatten_select_expressions(node.value)
+        return self.generic_visit(node)
+
+    def visit_SelectExpression(self, node: FTL.SelectExpression):
+        if is_plural_expression(node):
+            create_locale_plural_variants(node, self.locale)
+            NotranslateWrapper(
+                self.locale, self.replacements, node.selector
+            ).generic_visit(node)
+        return self.generic_visit(node)
+
+
+class ApplyPretranslation(Transformer):
+    """
+    During `visit()`, calls `callback(raw, wrapped, locale) -> (translation, service)` for each pattern.
+    """
+
+    def __init__(
+        self,
+        locale: Locale,
+        entry: FTL.EntryType,
+        callback: Callable[[str, str, str], tuple[Optional[str], str]],
+    ):
+        prep = PreparePretranslation(locale)
+        prep.visit(entry)
+        self.callback = callback
+        self.locale = locale
+        self.notranslate = NotranslateWrapper(locale, prep.replacements)
+        self.replacements = prep.replacements
+        self.services: list[str] = []
+
+    def applyReplacements(self, translation: str):
+        return re.sub(
+            r'<span id="pt-(\d+)" translate="no">[^<>]*</span>',
+            lambda m: self.replacements[int(m.group(1))],
+            translation,
+        )
+
+    def visit_Pattern(self, node: FTL.Pattern):
+        has_selects = False
+        source = ""
+        for el in node.elements:
+            if isinstance(el, FTL.TextElement):
+                source += el.value
+            elif isinstance(el.expression, FTL.SelectExpression):
+                self.generic_visit(el.expression)
+                has_selects = True
+            else:
+                text = self.notranslate.visit_Placeable(el)
+                source += text.value
+        if not has_selects and source != "":
+            # Machine translation treats each line as a separate sentence,
+            # hence we replace newline characters with spaces.
+            source = source.replace("\n", " ")
```

Comment:
Hm, not that I wrote this code, but... Should we instead move this to `get_google_translate_data()`? That way it would also work for Machinery and the generic engine.

Reply:
That would make sense as a separate change from this PR. Could also improve this by doing one or two of the following:
1. Collapse all whitespace sequences into a single space.
2. Leave newlines in the message if immediately preceded by a `.`